File: java/sfo_63_the_maximum_profit_of_stocks_s1/sfo_63_the_maximum_profit_of_stocks_s1.java
/*
* File: sfo_63_the_maximum_profit_of_stocks_s1.java
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

package sfo_63_the_maximum_profit_of_stocks_s1;

import include.*;
import java.util.*;

// ===== Solution Code =====
class Solution {
    public int maxProfit(int[] prices) {
        int cost = Integer.MAX_VALUE, profit = 0;
        for(int price : prices) {
            cost = Math.min(cost, price);
            profit = Math.max(profit, price - cost);
        }
        return profit;
    }
}

public class sfo_63_the_maximum_profit_of_stocks_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        int[] prices = { 7, 1, 5, 3, 6, 4 };
        // ====== Driver Code ======
        Solution slt = new Solution();
        int res = slt.maxProfit(prices);
        System.out.println(res);
    }
}

File: python/sfo_63_the_maximum_profit_of_stocks_s1.py
'''
File: sfo_63_the_maximum_profit_of_stocks_s1.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# ===== Solution Code =====
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        cost, profit = float("+inf"), 0
        for price in prices:
            cost = min(cost, price)
            profit = max(profit, price - cost)
        return profit

# ======= Test Case =======
prices = [7, 1, 5, 3, 6, 4]
# ====== Driver Code ======
slt = Solution()
res = slt.maxProfit(prices)
print(res)
